File: v1/src/main/java/com/api/v1/customers/services/CustomerRegistrationServiceImpl.java
package com.api.v1.customers.services;

import com.api.v1.customers.domain.CustomerEntity;
import com.api.v1.customers.domain.CustomerRepository;
import com.api.v1.customers.dtos.CustomerResponseDto;
import com.api.v1.customers.utils.CustomerResponseMapperUtil;
import com.api.v1.persons.dtos.PersonRegistrationRequestDto;
import com.api.v1.persons.services.PersonRegistrationService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class CustomerRegistrationServiceImpl implements CustomerRegistrationService {

    @Autowired
    private PersonRegistrationService personRegistrationService;

    @Autowired
    private CustomerRepository customerRepository;

    @Override
    public Mono<CustomerResponseDto> register(@Valid PersonRegistrationRequestDto requestDto) {
        return personRegistrationService
                .register(requestDto)
                .flatMap(person -> customerRepository
                        .save(CustomerEntity.of(person))
                        .flatMap(customer -> Mono.just(CustomerResponseMapperUtil.map(customer)))
                );
    }

}

File: kotlin/v3/src/main/kotlin/com/api/v3/customers/services/CustomerRegistrationServiceImpl.kt
package com.api.v3.customers.services

import com.api.v3.customers.domain.Customer
import com.api.v3.customers.domain.CustomerRepository
import com.api.v3.customers.dtos.CustomerResponseDto
import com.api.v3.customers.utils.CustomerResponseMapper
import com.api.v3.persons.dtos.PersonRegistrationRequestDto
import com.api.v3.persons.services.PersonRegistrationService
import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.stereotype.Service

@Service
internal class CustomerRegistrationServiceImpl(
    private val personRegistrationService: PersonRegistrationService,
    private val customerRepository: CustomerRepository
): CustomerRegistrationService {

    override suspend fun register(requestDto: @Valid PersonRegistrationRequestDto): CustomerResponseDto {
        return withContext(Dispatchers.IO) {
            val savedPerson = personRegistrationService.register(requestDto)
            val customer = Customer.of(savedPerson)
            val savedCustomer = customerRepository.save(customer)
            CustomerResponseMapper.map(savedCustomer)
        }
    }

}